File: backend/app/schemas/apm.py
from __future__ import annotations

from datetime import datetime
from pydantic import BaseModel


# ── Business Transactions ─────────────────────────────────────────

class BusinessTransactionPayload(BaseModel):
    app_name: str
    endpoint: str
    method: str
    status_code: int
    response_time_ms: float
    is_error: int = 0
    error_message: str | None = None
    trace_id: str | None = None
    user_id: str | None = None
    collected_at: datetime


class BusinessTransactionItem(BaseModel):
    id: int
    host_id: int | None
    app_name: str
    endpoint: str
    method: str
    status_code: int
    response_time_ms: float
    is_error: int
    error_message: str | None
    trace_id: str | None
    collected_at: datetime

    class Config:
        from_attributes = True


class BtSummary(BaseModel):
    app_name: str
    total_requests: int
    avg_response_time_ms: float
    error_count: int
    error_rate: float
    p95_response_time_ms: float | None = None


class BtAgentPayload(BaseModel):
    """Batch of business transactions from an agent."""
    hostname: str
    transactions: list[BusinessTransactionPayload]


class BtAgentResponse(BaseModel):
    success: bool
    message: str
    count: int


# ── Distributed Tracing ──────────────────────────────────────────

class SpanPayload(BaseModel):
    span_id: str
    parent_span_id: str | None = None
    service_name: str
    operation_name: str
    span_kind: str = "internal"
    status: str = "ok"
    duration_ms: float
    started_at: datetime
    attributes: str | None = None  # JSON string
    events: str | None = None  # JSON string


class TracePayload(BaseModel):
    trace_id: str
    root_service: str
    root_endpoint: str | None = None
    root_method: str | None = None
    status_code: int | None = None
    total_duration_ms: float
    has_error: int = 0
    started_at: datetime
    spans: list[SpanPayload]


class SpanItem(BaseModel):
    id: int
    trace_id: str
    span_id: str
    parent_span_id: str | None
    service_name: str
    operation_name: str
    span_kind: str
    status: str
    duration_ms: float
    started_at: datetime
    attributes: str | None
    events: str | None

    class Config:
        from_attributes = True


class TraceListItem(BaseModel):
    id: int
    trace_id: str
    root_service: str
    root_endpoint: str | None
    root_method: str | None
    status_code: int | None
    total_duration_ms: float
    span_count: int
    has_error: int
    started_at: datetime

    class Config:
        from_attributes = True


class TraceDetail(TraceListItem):
    spans: list[SpanItem]


class TraceAgentPayload(BaseModel):
    """Batch of traces from an agent."""
    hostname: str
    traces: list[TracePayload]


class TraceAgentResponse(BaseModel):
    success: bool
    message: str
    count: int


# ── Baselines & Anomalies ────────────────────────────────────────

class BaselineItem(BaseModel):
    id: int
    host_id: int
    metric_name: str
    mean: float
    stddev: float
    min_val: float
    max_val: float
    sample_count: int
    window_hours: int
    computed_at: datetime

    class Config:
        from_attributes = True


class AnomalyItem(BaseModel):
    id: int
    host_id: int
    hostname: str | None = None
    metric_name: str
    observed_value: float
    baseline_mean: float
    baseline_stddev: float
    deviation_sigma: float
    severity: str
    status: str
    detected_at: datetime
    resolved_at: datetime | None

    class Config:
        from_attributes = True


class AnomalySummary(BaseModel):
    total_anomalies: int
    open_anomalies: int
    warning_count: int
    critical_count: int


# ── Log Analytics ─────────────────────────────────────────────────

class LogPayload(BaseModel):
    source: str
    level: str
    message: str
    trace_id: str | None = None
    logged_at: datetime


class LogAgentPayload(BaseModel):
    hostname: str
    logs: list[LogPayload]


class LogAgentResponse(BaseModel):
    success: bool
    message: str
    count: int


class LogItem(BaseModel):
    id: int
    host_id: int
    hostname: str
    source: str
    level: str
    message: str
    trace_id: str | None
    logged_at: datetime

    class Config:
        from_attributes = True


class LogSearchResult(BaseModel):
    total: int
    logs: list[LogItem]


class LogSummary(BaseModel):
    total_logs: int
    error_count: int
    warn_count: int
    info_count: int


# ── Service Topology ─────────────────────────────────────────────

class ServiceNodeItem(BaseModel):
    id: int
    service_name: str
    service_type: str
    host_id: int | None
    status: str
    avg_response_time_ms: float | None
    request_rate: float | None
    error_rate: float | None
    last_seen_at: datetime | None

    class Config:
        from_attributes = True


class ServiceDependencyItem(BaseModel):
    id: int
    source_service: str
    target_service: str
    call_count: int
    error_count: int
    avg_duration_ms: float | None
    last_seen_at: datetime

    class Config:
        from_attributes = True


class TopologyGraph(BaseModel):
    nodes: list[ServiceNodeItem]
    edges: list[ServiceDependencyItem]


# ── Code-Level Diagnostics ───────────────────────────────────────

class DiagnosticSnapshotItem(BaseModel):
    id: int
    host_id: int | None
    app_name: str
    snapshot_type: str
    duration_seconds: float | None
    top_functions: str | None  # JSON
    memory_summary: str | None  # JSON
    thread_dump: str | None  # JSON
    triggered_by: str | None
    collected_at: datetime

    class Config:
        from_attributes = True


class DiagnosticPayload(BaseModel):
    hostname: str
    app_name: str
    snapshot_type: str
    duration_seconds: float | None = None
    top_functions: str | None = None
    memory_summary: str | None = None
    thread_dump: str | None = None
    triggered_by: str | None = None
    collected_at: datetime


class DiagnosticResponse(BaseModel):
    success: bool
    message: str
    snapshot_id: int

File: frontend/src/types/apm.ts
/* ── Business Transactions ──────────────────────────────── */

export interface BusinessTransactionItem {
  id: number;
  host_id: number | null;
  app_name: string;
  endpoint: string;
  method: string;
  status_code: number;
  response_time_ms: number;
  is_error: number;
  error_message: string | null;
  trace_id: string | null;
  collected_at: string;
}

export interface BtSummary {
  app_name: string;
  total_requests: number;
  avg_response_time_ms: number;
  error_count: number;
  error_rate: number;
  p95_response_time_ms: number | null;
}

/* ── Distributed Tracing ───────────────────────────────── */

export interface TraceListItem {
  id: number;
  trace_id: string;
  root_service: string;
  root_endpoint: string | null;
  root_method: string | null;
  status_code: number | null;
  total_duration_ms: number;
  span_count: number;
  has_error: number;
  started_at: string;
}

export interface SpanItem {
  id: number;
  trace_id: string;
  span_id: string;
  parent_span_id: string | null;
  service_name: string;
  operation_name: string;
  span_kind: string;
  status: string;
  duration_ms: number;
  started_at: string;
  attributes: string | null;
  events: string | null;
}

export interface TraceDetail extends TraceListItem {
  spans: SpanItem[];
}

/* ── Baselines & Anomalies ─────────────────────────────── */

export interface BaselineItem {
  id: number;
  host_id: number;
  metric_name: string;
  mean: number;
  stddev: number;
  min_val: number;
  max_val: number;
  sample_count: number;
  window_hours: number;
  computed_at: string;
}

export interface AnomalyItem {
  id: number;
  host_id: number;
  hostname: string | null;
  metric_name: string;
  observed_value: number;
  baseline_mean: number;
  baseline_stddev: number;
  deviation_sigma: number;
  severity: string;
  status: string;
  detected_at: string;
  resolved_at: string | null;
}

export interface AnomalySummary {
  total_anomalies: number;
  open_anomalies: number;
  warning_count: number;
  critical_count: number;
}

/* ── Log Analytics ─────────────────────────────────────── */

export interface LogItem {
  id: number;
  host_id: number;
  hostname: string;
  source: string;
  level: string;
  message: string;
  trace_id: string | null;
  logged_at: string;
}

export interface LogSearchResult {
  total: number;
  logs: LogItem[];
}

export interface LogSummary {
  total_logs: number;
  error_count: number;
  warn_count: number;
  info_count: number;
}

/* ── Service Topology ──────────────────────────────────── */

export interface ServiceNodeItem {
  id: number;
  service_name: string;
  service_type: string;
  host_id: number | null;
  status: string;
  avg_response_time_ms: number | null;
  request_rate: number | null;
  error_rate: number | null;
  last_seen_at: string | null;
}

export interface ServiceDependencyItem {
  id: number;
  source_service: string;
  target_service: string;
  call_count: number;
  error_count: number;
  avg_duration_ms: number | null;
  last_seen_at: string;
}

export interface TopologyGraph {
  nodes: ServiceNodeItem[];
  edges: ServiceDependencyItem[];
}

/* ── Code-Level Diagnostics ────────────────────────────── */

export interface DiagnosticSnapshotItem {
  id: number;
  host_id: number | null;
  app_name: string;
  snapshot_type: string;
  duration_seconds: number | null;
  top_functions: string | null;
  memory_summary: string | null;
  thread_dump: string | null;
  triggered_by: string | null;
  collected_at: string;
}
